### PR TITLE
Fix regression with preserving `this` context

### PR DESCRIPTION
**Pull request by darcyparker**
See https://github.com/sindresorhus/p-debounce/pull/12#issuecomment-762918962

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -9,7 +9,17 @@ const pDebounce = (fn, wait, options = {}) => {
...
 	let resolveList = [];
 
+	// IMPORTANT: Must preserve calling context that debounced fn is called
+	// * Returned debounced fn is written as `function() { }` so that its called `this` context
+	//   is known.
+	// * `new Promise(resolve => {})` and setTimeout's callbacks are arrow fn(), so `context` is
+	//   not required for ES6.  But if pDebounce is transpiled to ES5, `context` is required because
+	//   the arrow fn will become a `function() {...}`
+	// * `result` MUST be calculated by calling fn() in same context that pDebounce()'s returned
+	//   function is called.  Hence `fn.apply(context, arguments_)` rather than `fn(...arguments_)`
```
Comment (fregante) [inline index.js:19]: I think this comment can be replaced by something like `Preserve the context when the debounced function is called on this`

Comment (darcyparker) [reply index.js]: Thank you.  Simplifying the comment is helpful.  I added a commit that is similar to your suggestion.

File: index.js [hunk 2]
```diff
@@ -40,13 +43,15 @@ const pDebounce = (fn, wait, options = {}) => {
...
+	// Also see notes in pDebounce() about preserving calling context
+	return async function (...arguments_) {
 		if (currentPromise) {
 			return currentPromise;
 		}
 
+		const context = this;
 		try {
-			currentPromise = function_(...arguments_);
+			currentPromise = function_.apply(context, arguments_);
```
Comment (sindresorhus) [inline index.js:54]:
```suggestion
			currentPromise = function_.apply(this, arguments_);
```

Comment (darcyparker) [reply index.js]: Sure - I made this change.  I understand you aren't interested in a change that makes the code better for transpilation. https://github.com/sindresorhus/p-throttle/pull/25#discussion_r561542921

File: index.js (hunk 1 above)
Comment (darcyparker) [reply index.js]: The comment is now removed given https://github.com/sindresorhus/p-debounce/pull/16#discussion_r561587514

File: test.js [hunk 3]
```diff
@@ -79,3 +79,41 @@ test('leading option - does not call input function after timeout', async t => {
...
+		// If `this` is not preserved by `pDebounce()` or `pDebounce.promise()`,
+		// then `this` will be undefined.
+		return this;
+	}
+};
+
+[
+	['pDebounce()', pDebounce],
+	['pDebounce().promise()', pDebounce.promise]
+].forEach(([name, debounceFn]) => {
```
Comment (sindresorhus) [inline test.js:107]: Move the array to a variable and then use a `for-of` loop. Would make it a bit more readable.

Comment (darcyparker) [reply test.js]: done

**Comment by sindresorhus** [comment]
Thanks for fixing this :)

**Pull request merged**